### PR TITLE
feature/optimistic recovery

### PR DESCRIPTION
**Pull request by colinsurprenant**
Optimistic recovery feature.

With this feature, on startup, the head page data file is opened using the new `PageIO.recover()` method which sequentially reads the data file from the start and try to find valid written elements regardless of the head checkpoint. This is possible thanks to the per-element checksums which allow self-integrity checking in the data file. 

### TODO

- [x] see about capacity as an it versus file size as long (see TODO in PageIO)
- [x] DRY code between open/recover methods
- [x] DRY code between MmapPageIO and ByteBufferPageIO to ease testing
- [x] add more unit tests
- [x] remove debug traces

### CODE REVIEW

**Comment by colinsurprenant** [comment]
It is a bit out of scope of this PR but I decided to extract/DRY code into a new abstract class `AbstractByteBufferPageIO` - since most code between `ByteBufferPageIO` and `MmapPageIO` was identical and in order to add tests for the `recover()` method I would have had to duplicate even more code so I decided to fix that now.

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        if (this.elementCount <= 0) {
+            this.minSeqNum = 0;
+        }
+    }
+
+    // read and vadidate next element at page head
+    // @param verifyChecksum if true the actual element data will be read + checksumed and compared to written checksum
+    private void readNextElement(long expectedSeqNum, boolean verifyChecksum) throws PageIOInvalidElementException {
+        // if there is no room for the seqNum and length bytes stop here
+        if (this.head + SEQNUM_SIZE + LENGTH_SIZE > capacity) { throw new PageIOInvalidElementException("cannot read seqNum and length bytes past buffer capacity"); }
```
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: "past buffer capacity" sounds strange to me. Maybe it's ok as worded, but it feels more like "We want to read the next element, but there are fewer than X bytes available before the end of the page, so this is probably a partial write that occurred when Logstash was last terminated abnormally" ?

Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: Or something like: "While recovering, we reached the end of the page while trying to read the next entry. This means ... something something partial write ..."

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+        if (seqNum != expectedSeqNum) { throw new PageIOInvalidElementException(String.format("Element seqNum %d is expected to be %d", seqNum, expectedSeqNum)); }
+
+        int length = buffer.getInt();
+        this.head += LENGTH_SIZE;
+
+        // length must be > 0
+        if (length <= 0) { throw new PageIOInvalidElementException("Element invalid length"); }
+
+        // if there is no room for the proposed data lenght and checksum just stop here
```
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: typo "length"

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+        long seqNum = buffer.getLong();
+        this.head += SEQNUM_SIZE;
+
+        if (seqNum != expectedSeqNum) { throw new PageIOInvalidElementException(String.format("Element seqNum %d is expected to be %d", seqNum, expectedSeqNum)); }
+
+        int length = buffer.getInt();
+        this.head += LENGTH_SIZE;
+
+        // length must be > 0
+        if (length <= 0) { throw new PageIOInvalidElementException("Element invalid length"); }
```
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: Do you want to check for too-large length also? (larger than page size?)

File: logstash-core/src/test/java/org/logstash/common/io/MemoryCheckpointTest.java [hunk 4]
```diff
@@ -39,7 +41,11 @@ public void writeNewReadExisting() throws IOException {
 
     @Test
     public void readInnexisting() throws IOException {
-        Checkpoint checkpoint = io.read("checkpoint.invalid");
-        assertThat(checkpoint, is(equalTo(null)));
+        try {
+            io.read("checkpoint.invalid");
+            fail("expected NoSuchFileException");
+        } catch (NoSuchFileException e) {
```
Comment (jordansissel) [inline logstash-core/src/test/java/org/logstash/common/io/MemoryCheckpointTest.java]: Testing suggestion:

* Change annotation to: `@Test(exception = NoSuchFileException)` (or whatever, I'm writing by hand here, not tested)
* Change function to be 1 line: `io.read("checkpoint.invalid")`

File: logstash-core/src/test/java/org/logstash/common/io/ByteBufferPageIOTest.java [hunk 5]
```diff
@@ -63,64 +64,217 @@ public void hasSpaceAfterWrite() throws IOException {
...
+            ByteBufferPageIO io = newEmptyPageIO();
+            byte[] bytes = element.serialize();
+            int checksum = io.checksum(bytes);
+            bytes[1] = 0x01;
+            io.write(bytes, 1L, bytes.length, checksum);
+            return io.dump();
+        }
+    );
+
+    private final List<BufferGenerator> doubleGenerators = Arrays.asList(
```
Comment (jordansissel) [inline logstash-core/src/test/java/org/logstash/common/io/ByteBufferPageIOTest.java]: This would probably be better written using JUnit parameterized tests for the generated dumps.

```java
@RunWith(Parameterized.class)
public class MyTestClass {
    @Parameters
    public static Collection<byte[]> data() {
       ... generate all `io.dump()` calls here and return a Collection<> of them.
    }

    @Parameter
    public byte[] bytes;

    @Test
    public void openInvalidElement() {
         ByteBufferPageIO io = newPageIO(bytes.length, bytes);
         io.open(1L, 1);

         assertThat(io.getElementCount(), is(equalTo(1)));
         assertThat(io.getMinSeqNum(), is(equalTo(1L))); 
    }
}
```

This has nice properties of avoiding the iteration you do on singleGenerators and doubleGenerators. You could also use this to simplify the invalid write generation.

File: logstash-core/src/test/java/org/logstash/ackedqueue/TestSettings.java [hunk 6]
```diff
@@ -9,7 +9,7 @@
 
 public class TestSettings {
 
-    public static Settings getSettings(int capacity) {
+    public static Settings volatileQueueSettings(int capacity) {
```
Comment (jordansissel) [inline logstash-core/src/test/java/org/logstash/ackedqueue/TestSettings.java]: Why the rename?

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+            }
+        }
+    }
+
+    // recover will overwrite/update/set this object minSeqNum, capacity and elementCount attributes
+    // to reflect what it recovered from the page
+    @Override
+    public void recover() throws IOException {
+        getBuffer().position(0);
+        this.version = getBuffer().get();
```
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: The `version` value should be validated here. Is it a known/supported version?

File: logstash-core/src/test/java/org/logstash/common/io/MemoryCheckpointTest.java (hunk 4 above)
Comment (colinsurprenant) [inline logstash-core/src/test/java/org/logstash/common/io/MemoryCheckpointTest.java]: uh ... weird ... not sure why I did not use that in the first place really 🤔

File: logstash-core/src/test/java/org/logstash/common/io/ByteBufferPageIOTest.java (hunk 5 above)
Comment (colinsurprenant) [inline logstash-core/src/test/java/org/logstash/common/io/ByteBufferPageIOTest.java]: TIL @Parameter ... will look.

File: logstash-core/src/test/java/org/logstash/ackedqueue/TestSettings.java (hunk 6 above)
Comment (colinsurprenant) [inline logstash-core/src/test/java/org/logstash/ackedqueue/TestSettings.java]: because the `TestSettings` class had `getSettings()` and `getSettingsCheckpointFilePageMemory()` which has 2 problems:
1- since `TestSettings` now exposes both volatile and persisted settings, `getSettings()` means nothing.
2- I have not idea what `getSettingsCheckpointFilePageMemory()` even mean

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java (hunk 3 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: well, it is implicitly checked below with `if (this.head + length + CHECKSUM_SIZE > capacity)` ...

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java (hunk 2 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: +1

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java (hunk 1 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: we ... it's not necessarily while recovering, this check also happens on `open()` and is basically an integrity check and I have not idea how this can actually happen but it's just to make sure we can do the following get___ without triggering a buffer overflow. In this case I wouldn't try to infer what has caused this but prefer to simply report the fact ... but we can reword if it's not clear...

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java (hunk 7 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: yeah, no, we did not add validation on the version yet...

File: logstash-core/src/test/java/org/logstash/ackedqueue/TestSettings.java (hunk 6 above)
Comment (jordansissel) [inline logstash-core/src/test/java/org/logstash/ackedqueue/TestSettings.java]: > getSettingsCheckpointFilePageMemory

I kind of want to name all the methods this, now. haha.

Thanks for the clarification <3

Comment (colinsurprenant) [inline logstash-core/src/test/java/org/logstash/ackedqueue/TestSettings.java]: lol - the real story is, I was reading some tests around `Queue.open()` and stumbled on `getSettingsCheckpointFilePageMemory()` and thought WTF is this setting?!?!  I seriously had NO idea what this settings was about :P

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java (hunk 1 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: not sure how to reword.

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java (hunk 3 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: so a fiew check is made to make sure length is greater than zero, and then if the length is actually valid with respect to the buffer size:
```
        // length must be > 0
        if (length <= 0) { throw new PageIOInvalidElementException("Element invalid length"); }

        // if there is no room for the proposed data lenght and checksum just stop here
        if (this.head + length + CHECKSUM_SIZE > capacity) { throw new PageIOInvalidElementException("cannot read element payload and checksum past buffer capacity"); }
```

is that ok with you?

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java (hunk 7 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: fixed

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java (hunk 3 above)
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: ++

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java (hunk 1 above)
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: I'm not sure how to reword it at this time, either.

How about adding a TODO comment above indicating that we know this isn't a great exception message but at the time of writing we couldn't come up with anything better.

File: logstash-core/src/test/java/org/logstash/ackedqueue/TestSettings.java (hunk 6 above)
Comment (jordansissel) [inline logstash-core/src/test/java/org/logstash/ackedqueue/TestSettings.java]: haha.

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java [hunk 8]
```diff
@@ -162,33 +163,45 @@ public void open() throws IOException {
         for (int pageNum = headCheckpoint.getFirstUnackedPageNum(); pageNum < headCheckpoint.getPageNum(); pageNum++) {
```
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: shouldn't this `headCheckpoint.getFirstUnackedPageNum()` happen after you optimistically recover?

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java [hunk 9]
```diff
@@ -223,30 +238,30 @@ private void add(Checkpoint checkpoint, PageIO pageIO) throws IOException {
             } else {
                 // create a tail page with a null PageIO and add it to tail pages but not unreadTailPages
                 // since it is fully read because also fully acked
+                // TODO: I don't like this null pageIO tail page...
```
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: neither do I

File: logstash-core/src/main/java/org/logstash/ackedqueue/TailPage.java [hunk 10]
```diff
@@ -14,19 +14,14 @@ public TailPage(HeadPage page) {
     }
 
     // create a new TailPage object for an exiting Checkpoint and data file
+    // @param @pageIO is expected to be open/recover/create
```
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/ackedqueue/TailPage.java]: can we use proper javadoc linking syntax here?

File: logstash-core/src/main/java/org/logstash/ackedqueue/TailPage.java [hunk 11]
```diff
@@ -14,19 +14,14 @@ public TailPage(HeadPage page) {
...
         super(checkpoint.getPageNum(), queue, checkpoint.getMinSeqNum(), checkpoint.getElementCount(), checkpoint.getFirstUnackedSeqNum(), new BitSet(), pageIO);
 
         // this page ackedSeqNums bitset is a new empty bitset, if we have some acked elements, set them in the bitset
         if (checkpoint.getFirstUnackedSeqNum() > checkpoint.getMinSeqNum()) {
             this.ackedSeqNums.flip(0, (int) (checkpoint.getFirstUnackedSeqNum() - checkpoint.getMinSeqNum()));
         }
-
-        if (pageIO != null) {
-            // open the data file and reconstruct the IO object internal state
-            pageIO.open(checkpoint.getMinSeqNum(), checkpoint.getElementCount());
```
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/ackedqueue/TailPage.java]: the only reason we are safe in removing this is because this is the only place we construct?

https://github.com/elastic/logstash/pull/6519/files#diff-69b2c41a7ba6596afa233505525212bbR169

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java (hunk 1 above)
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: nitpick, can we expand this from being a one-liner?

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java [hunk 12]
```diff
@@ -2,6 +2,7 @@
 
 import org.logstash.ackedqueue.Queueable;
 import org.logstash.ackedqueue.SequencedList;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
```
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: this doesn't seem to be used. why not have this be thrown for all of those `/** zilch **/` spaces?

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java [hunk 13]
```diff
@@ -41,208 +24,41 @@ public ByteBufferPageIO(int capacity) throws IOException {
...
     @Override
-    public int persistedByteCount(int byteCount) {
-        return ByteBufferPageIO._persistedByteCount(byteCount);
-    }
-
-    private long maxSeqNum() {
-        return this.minSeqNum + this.elementCount - 1;
-    }
-
+    protected ByteBuffer getBuffer() { return this.buffer; }
```
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: expand to multiline

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java [hunk 14]
```diff
@@ -41,208 +24,41 @@ public ByteBufferPageIO(int capacity) throws IOException {
...
 
     // below public methods only used by tests
 
-    public int getWritePosition() {
-        return this.head;
-    }
+    // TODO: static method for tests - should refactor
+    public static int _persistedByteCount(int byteCount) { return SEQNUM_SIZE + LENGTH_SIZE + byteCount + CHECKSUM_SIZE; }
```
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: expand to multiline.

can we remove the `_` part? and if it is only used in tests, we can make this package-private.

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java [hunk 15]
```diff
@@ -41,208 +24,41 @@ public ByteBufferPageIO(int capacity) throws IOException {
...
 
-    public int getElementCount() {
-        return this.elementCount;
-    }
+    public int getWritePosition() { return this.head; }
```
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: multiline

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java [hunk 16]
```diff
@@ -41,208 +24,41 @@ public ByteBufferPageIO(int capacity) throws IOException {
...
 
-    public byte[] dump() {
-        return this.buffer.array();
-    }
+    public byte[] dump() { return this.buffer.array(); }
```
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: multiline

File: logstash-core/src/main/java/org/logstash/common/io/MmapPageIO.java [hunk 17]
```diff
@@ -1,199 +1,74 @@
...
+    @Override
+    public void recover() throws IOException {
+        mapFile(!STRICT_CAPACITY);
+        super.recover();
+    }
 
-                this.offsetMap.add(head);
-                this.head += SEQNUM_SIZE;
+    // memory map data file to this.buffer and read initial version byte
+    // @param strictCapacity if true verify that data file size is same as confgured page capcity, if false update page capcity to actual file size
```
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/MmapPageIO.java]: proper javadoc format here?

File: logstash-core/src/main/java/org/logstash/common/io/PageIO.java [hunk 18]
```diff
@@ -10,10 +10,20 @@
 public interface PageIO extends Closeable {
 
     // the concrete class should be constructed with the pageNum, capacity and dirPath attributes
+    // and open/recover/create must first be called to setup with physical data file
+    //
+    // TODO: we should probably refactor this with a factory to force the creation of a fully
```
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/PageIO.java]: 👍 . why not do this for this PR?

File: logstash-core/src/main/java/org/logstash/common/io/PageIO.java [hunk 19]
```diff
@@ -10,10 +10,20 @@
...
+    //       initialized concrete object with either open/recover/create instead of allowing
+    //       a partially initialized object using the concrete class constructor. Not sure.
 
     // open an existing data container and reconstruct internal state if required
     void open(long minSeqNum, int elementCount) throws IOException;
 
+    // optimistically recover an existing data container and reconstruct internal state
+    // with the actual read/recovered data. this is only useful il reading a head page
```
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/PageIO.java]: proper javadoc? `s/il/if`

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+                readNextElement(this.minSeqNum + i, VERIFY_CHECKSUM);
+                this.elementCount += 1;
+            } catch (PageIOInvalidElementException e) {
+                // simply stop at first invalid element
+                break;
+            }
+        }
+
+        // if we were not able to read any element just reset minSeqNum to zero
+        if (this.elementCount <= 0) {
```
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: the `<` part  may not be necessary here. Since we only increment elementCount and it starts at 0 above.

Maybe use an assert to crash if elementCount < 0, then make this check be `if (this.elementCount == 0)` instead of `<= 0` ?

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+        this.minSeqNum = 0L;
+        this.elementCount = 0;
+    }
+
+    @Override
+    public void write(byte[] bytes, long seqNum) throws IOException {
+        write(bytes, seqNum, bytes.length, checksum(bytes));
+    }
+
+    protected int write(byte[] bytes, long seqNum, int length, int checksum) {
```
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: Possibly out of scope for this change, but since `bytes.length` is known, does this method need to take a `length` parameter?

File: logstash-core/src/main/java/org/logstash/common/io/wip/MemoryPageIOStream.java [hunk 22]
```diff
@@ -70,6 +71,11 @@ public MemoryPageIOStream(int capacity, byte[] initialBytes) throws IOException
     }
 
     @Override
+    public void recover() throws IOException {
+        throw new NotImplementedException();
```
Comment (jordansissel) [inline logstash-core/src/main/java/org/logstash/common/io/wip/MemoryPageIOStream.java]: Curiosity - Is this the right exception to throw, or should we make our own class for this?

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java (hunk 21 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: yes, because it allows forcing each value independently of the actual bytes content. note that this is not a public method, and it is used by tests to inject incorrect values.

File: logstash-core/src/main/java/org/logstash/common/io/wip/MemoryPageIOStream.java (hunk 22 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/wip/MemoryPageIOStream.java]: did not pay much attention to this, as it is "unused" code yet ("wip" namespace) and this is a dev-time place holder just to make sure it crashes if someone starts to work with it and run tests for example - just a friendly remainder it needs to be implemented. :P

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java (hunk 8 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: the optimistic recovery only applies to the head page elements.

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java (hunk 9 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: I will refactor/clean this but not in this PR

File: logstash-core/src/main/java/org/logstash/ackedqueue/TailPage.java (hunk 10 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/ackedqueue/TailPage.java]: oh - yeah - muscle memory :P

File: logstash-core/src/main/java/org/logstash/ackedqueue/TailPage.java (hunk 11 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/ackedqueue/TailPage.java]: I removed all PageIO opening from the *Page classes and a properly opened PageIO must be fed to the *Page

File: logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java (hunk 1 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/AbstractByteBufferPageIO.java]: I personally prefer it as a one-liner - I really like the fact that this validation code uses less visual real estate and makes the general flow more readable. Again, personal preference.

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java (hunk 12 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: will remove. 
the `/** zilch **/` should stay as-is since these are no-ops methods, not `NotImplementedException`s

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java (hunk 13 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: I do prefer this concise style in such cases.

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java (hunk 14 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: you mean `protected` ?

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java (hunk 15 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: I prefer it single line

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java (hunk 16 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: I prefer it single line

File: logstash-core/src/main/java/org/logstash/common/io/MmapPageIO.java (hunk 17 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/MmapPageIO.java]: absolutely :P

File: logstash-core/src/main/java/org/logstash/common/io/PageIO.java (hunk 18 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/PageIO.java]: well, it is not really related to this PR and non-critical - the other refactors were triggered by extremely non-DRY code which I think was justified while working on this. I think we should open a followup issue

File: logstash-core/src/main/java/org/logstash/common/io/PageIO.java (hunk 19 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/PageIO.java]: yup

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java (hunk 8 above)
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: hmm, I see. so if the head checkpoint is old and this value is incorrect, it doesn't matter

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java (hunk 9 above)
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: ok

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java (hunk 12 above)
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: got it

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java (hunk 13 above)
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: ok

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java (hunk 14 above)
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: no, just no `public`. just `static int`. will allow tests to see it since I assume they are in `org.logstash.common.io`

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java (hunk 15 above)
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: ok

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java (hunk 16 above)
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: ok

File: logstash-core/src/main/java/org/logstash/common/io/PageIO.java (hunk 18 above)
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/PageIO.java]: oh, ok. I just didn't see this TODO copy/pasted from anywhere, so I assumed this was a related note

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java (hunk 8 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: well ... the head checkpoint is the only way to retrieve the `firstUnackedPageNum` so we have no choice to rely on it. I guess the worst case would be that `firstUnackedPageNum` points to a page that is now fully acked which should not be a problem (need to verify) but if that page has been purged it might be problematic. I will open a followup issue to verify this edge case.

Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: followup in #6592

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java (hunk 9 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: followup in #6593

File: logstash-core/src/main/java/org/logstash/ackedqueue/TailPage.java (hunk 10 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/ackedqueue/TailPage.java]: fixed

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java (hunk 14 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: so - at this point it cannot "easily" be other than public because there are many usage outside that package ... we should defer that to another issue/pr as this is really out of scope for this pr. ok?

Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: followup in #6594

File: logstash-core/src/main/java/org/logstash/common/io/MmapPageIO.java (hunk 17 above)
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/MmapPageIO.java]: oh you mean using `/** */` ?
I see I have been using `//` type of comments everywhere and this is not javadoc compatible, I will need to do a global cleanup.

Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/common/io/MmapPageIO.java]: followup in #6595

**Comment by colinsurprenant** [comment]
@jordansissel @talevy thanks for the reviews/comments - just push a few commits and followups.

File: logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java (hunk 8 above)
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java]: hmm. after trying to force this to become an issue... I can't. Since purging of tail files only happens on ack... a checkpoint would also be created. There would only be an issue if something killed the process.. in which case, anything can be corrupt

File: logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java (hunk 12 above)
Comment (talevy) [inline logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java]: still not removed?

**Comment by jordansissel** [comment]
LGTM

**Comment by colinsurprenant** [comment]
rebased + squashed and merged in master and 5.x

**Comment by jsvd** [comment]
for future reference the commit on master for this PR is https://github.com/elastic/logstash/commit/356dd71e1e69b9fb652757cf0a4b72b89eca3248
